### PR TITLE
Issue #4: timestamp field

### PR DESCRIPTION
**Pull request by idrissneumann**

### CODE REVIEW

**Comment by idrissneumann** [comment]
At this stage, seems ok:

Screenshot 2023-12-14 at 10 46 04

Will test if there's some regression on the search

**Comment by idrissneumann** [comment]
Regressions tests ok:

Screenshot 2023-12-14 at 11 07 52

**Comment by fmassot** [comment]
I get a 500 if the index does not exist
Screenshot from 2023-12-15 15-33-32

**Comment by idrissneumann** [comment]
> I get a 500 if the index does not exist !

@fmassot It's fixed with ae3d633256cc24bce0a2f6f303e41b77030adc35

Screenshot 2023-12-16 at 13 55 16

File: pkg/quickwit/quickwit.go [hunk 1]
```diff
@@ -26,6 +26,88 @@ type QuickwitDatasource struct {
 	dsInfo es.DatasourceInfo
 }
 
+type QuickwitMapping struct {
+	IndexConfig struct {
+		DocMapping struct {
+			TimestampField string `json:"timestamp_field"`
+			FieldMappings  []struct {
+				Name         string   `json:"name"`
+				InputFormats []string `json:"input_formats"`
```
Comment (fmassot) [inline pkg/quickwit/quickwit.go:35]: we want the `output_format`
the struct is a bit more complex, you can check the typescript code, particularly the `FieldMapping` type and the  `getAllFields` function.
Currently I still get a 500 with the log

```
grafana-quickwit-datasource  | logger=context userId=0 orgId=1 uname= t=2023-12-16T21:44:38.439604142Z level=error msg="Internal server error" error="[plugin.downstreamError] client: failed to call resources: failed to receive call resource response: rpc error: code = Unknown desc = {\"message\":\"Unmarshalling body error: err = json: cannot unmarshal array into Go value of type quickwit.QuickwitMapping, body = [\\n  {\\...
```

Comment (idrissneumann) [reply pkg/quickwit/quickwit.go]: Fixed: I use the `output_format` instead and flag it as optional.

So even if it doesn't exists, you shouldn't get unmarshalling errors anymore.

Tested with otel mapping:

Screenshot 2023-12-18 at 17 11 41

The logs:

```shell
grafana-quickwit-datasource  | logger=plugin.quickwit-quickwit-datasource t=2023-12-18T16:11:34.091637818Z level=info msg="Found timestampFieldName = timestamp_nanos, timestampFieldFormat = unix_timestamp_nanos"
```

Comment (idrissneumann) [reply pkg/quickwit/quickwit.go]: Add recursivity for fields with an `object` type. No regression with the otel and yoi indices mappings:

Screenshot 2023-12-22 at 15 07 50


```shell
grafana-quickwit-datasource  | logger=plugin.quickwit-quickwit-datasource t=2023-12-22T14:06:07.857021125Z level=info msg="Calling quickwit endpoint: http://host.docker.internal:7280/api/v1/indexes/otel-logs-v0_6"
grafana-quickwit-datasource  | logger=plugin.quickwit-quickwit-datasource t=2023-12-22T14:06:08.274010024Z level=info msg="Found timestampFieldName = timestamp_nanos, timestampFieldFormat = unix_timestamp_nanos"
```

Comment (idrissneumann) [reply pkg/quickwit/quickwit.go]: Test with this mapping:

```json
{
  "doc_mapping": {
    "dynamic_mapping": {
      "description": "the description",
      "expand_dots": true,
      "fast": true,
      "indexed": true,
      "record": "basic",
      "stored": true,
      "tokenizer": "raw"
    },
    "field_mappings": [
      {
        "name": "yo",
        "type": "object",
        "field_mappings": [{
          "name": "timestamp",
          "type": "datetime",
          "fast": true
        }]
      }
    ],
    "max_num_partitions": 1,
    "mode": "dynamic",
    "partition_key": "yo",
    "store_source": true,
    "tag_fields": [],
    "timestamp_field": "yo.timestamp"
  },
  "index_id": "yoii",
  "search_settings": {
    "default_search_fields": [
      "yo"
    ]
  },
  "version": "0.6"
}
```

And those documents:

```json
{"yo":{"timestamp": 1703255181}}
{"yo":{"timestamp": 1703255200}}
```

KO, I'll have to concatenate with the parent field name:

Screenshot 2023-12-22 at 15 27 32

Comment (idrissneumann) [reply pkg/quickwit/quickwit.go]: Fixed:

Screenshot 2023-12-22 at 15 33 47

```shell
grafana-quickwit-datasource  | logger=plugin.quickwit-quickwit-datasource t=2023-12-22T14:33:37.941152856Z level=info msg="Calling quickwit endpoint: http://host.docker.internal:7280/api/v1/indexes/yoii"
grafana-quickwit-datasource  | logger=plugin.quickwit-quickwit-datasource t=2023-12-22T14:33:38.371661484Z level=info msg="Found timestampFieldName = yo.timestamp, timestampFieldFormat = rfc3339"
```

File: src/datasource.ts [hunk 2]
```diff
@@ -147,21 +178,19 @@ export class QuickwitDataSource
...
-    if (timestampField.field_mapping.output_format !== this.timeOutputFormat) {
-      return `Timestamp output format is declared as '${timestampField.field_mapping.output_format}' in the doc mapping, not '${this.timeOutputFormat}'.`;
-    }
+
+    let timeOutputFormat = timestampField.field_mapping.output_format || 'unknown';
     const supportedTimestampOutputFormats = ['unix_timestamp_secs', 'unix_timestamp_millis', 'unix_timestamp_micros', 'unix_timestamp_nanos', 'iso8601', 'rfc3339'];
-    if (!supportedTimestampOutputFormats.includes(this.timeOutputFormat)) {
-      return `Timestamp output format '${this.timeOutputFormat} is not yet supported.`;
+    if (!supportedTimestampOutputFormats.includes(timeOutputFormat)) {
+      return `Timestamp output format '${timeOutputFormat} is not yet supported.`;
```
Comment (fmassot) [inline src/datasource.ts:193]: great!

**Review by fmassot — CHANGES_REQUESTED** [review]
I think we're good, I did not find issues while testing it.

Can we add 1 or 2 tests on the backend side? This will really help to sleep at night :)

**Comment by idrissneumann** [comment]
> Can we add a few tests on the backend side? This will really help to sleep at night :)

It's done :)

**Comment by idrissneumann** [comment]
Last status check, still working great:

Screenshot 2023-12-28 at 13 11 44

```shell
grafana-quickwit-datasource  | logger=plugin.quickwit-quickwit-datasource t=2023-12-28T12:11:07.21707582Z level=info msg="Calling quickwit endpoint: http://host.docker.internal:7280/api/v1/indexes/yoii"
grafana-quickwit-datasource  | logger=plugin.quickwit-quickwit-datasource t=2023-12-28T12:11:07.633917111Z level=info msg="Found timestampFieldName = yo.timestamp, timestampFieldFormat = rfc3339"
```

**Review by fmassot — APPROVED** [review]

**Pull request merged**